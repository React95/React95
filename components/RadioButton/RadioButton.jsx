import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import {
  radioChecked,
  radioCheckedDisabled,
  radioUnchecked,
  radioUncheckedDisabled,
} from './radioButtonStates';

const Icon = styled.span`
  width: 12px;
  height: 12px;

  content: '';
  display: inline-block;

  position: absolute;
  left: 0;
  top: 0;

  background-image: url(${radioUnchecked});
`;

const Text = styled.span`
  padding: 1px;
  user-select: none;

  position: absolute;
  top: 0;
  left: 18px;
`;

const Field = styled.input.attrs({
  type: 'radio',
})`
  margin: 0;
  opacity: 0;

  &:focus ~ ${Text}, &:active + ${Text} {
    border: 1px dotted;
    padding: 0;
  }

  &:checked + ${Icon} {
    background-image: url(${radioChecked});
  }

  &:disabled + ${Icon} {
    background-image: url(${radioUncheckedDisabled});
  }

  &:checked &:disabled + ${Icon} {
    background-image: url(${radioCheckedDisabled});
  }
`;

const Label = styled.label`
  position: relative;
  margin-bottom: 10px;

  ${props => props.disabled
    && `
    color: #868686;
    text-shadow: 0.5px 0.5px #d2d2d2;
  `}
`;

const RadioButton = ({
  children, onChange, disabled, ...props
}) => (
  <Label disabled={disabled}>
    <Field onChange={onChange} disabled={disabled} {...props} />
    <Icon />
    <Text>{children}</Text>
  </Label>
);

RadioButton.propTypes = {
  children: PropTypes.string,
  onChange: PropTypes.func,
  disabled: PropTypes.bool,
};

RadioButton.defaultProps = {
  children: '',
  onChange: () => {},
  disabled: false,
};

export default RadioButton;
