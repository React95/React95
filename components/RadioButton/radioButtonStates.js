import radioChecked from 'assets/pattern/radio-checked.png';
import radioCheckedDisabled from 'assets/pattern/radio-checked-disabled.png';
import radioUnchecked from 'assets/pattern/radio-unchecked.png';
import radioUncheckedDisabled from 'assets/pattern/radio-unchecked-disabled.png';

export {
  radioChecked,
  radioCheckedDisabled,
  radioUnchecked,
  radioUncheckedDisabled,
};
