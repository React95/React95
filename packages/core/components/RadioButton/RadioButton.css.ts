import {
  radioChecked,
  radioCheckedDisabled,
  radioUnchecked,
  radioUncheckedDisabled,
} from './radioButtonStates';
import { globalStyle, style } from '@vanilla-extract/css';
import { contract } from '../themes/contract.css';

export const icon = style({
  width: contract.space[12],
  height: contract.space[12],
  display: 'inline-block',
  position: 'absolute',
  left: contract.space[0],
  top: contract.space[0],
  backgroundImage: `url('${radioUnchecked}')`,
  imageRendering: 'pixelated',
});

export const text = style({
  padding: contract.space[1],
  userSelect: 'none',
  position: 'absolute',
  top: '0',
  left: contract.space[18],
});

export const label = style({
  position: 'relative',
  marginBottom: contract.space[10],
  display: 'block',
  selectors: {
    '&:has(:disabled)': {
      color: contract.colors.materialTextDisabled,
      textShadow: `0.5px 0.5px ${contract.colors.materialTextDisabledShadow}`,
    },
  },
});

export const field = style({
  margin: 0,
  opacity: 0,
});

globalStyle(`${field}:focus ~ ${text}, &:active + ${text}`, {
  borderWidth: contract.space[1],
  borderStyle: 'dotted',
  padding: 0,
});

globalStyle(`${field}:checked + ${icon}`, {
  backgroundImage: `url('${radioChecked}')`,
});

globalStyle(`${field}:disabled + ${icon}`, {
  backgroundImage: `url('${radioUncheckedDisabled}')`,
});

globalStyle(`${field}:checked:disabled + ${icon}`, {
  backgroundImage: `url('${radioCheckedDisabled}')`,
});
