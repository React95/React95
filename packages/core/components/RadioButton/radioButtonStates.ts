import radioChecked from '../shared/assets/pattern/radio-checked.png';
import radioCheckedDisabled from '../shared/assets/pattern/radio-checked-disabled.png';
import radioUnchecked from '../shared/assets/pattern/radio-unchecked.png';
import radioUncheckedDisabled from '../shared/assets/pattern/radio-unchecked-disabled.png';

export {
  radioChecked,
  radioCheckedDisabled,
  radioUnchecked,
  radioUncheckedDisabled,
};
