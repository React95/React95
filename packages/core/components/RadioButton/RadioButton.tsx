import React, { forwardRef } from 'react';
import { field, icon, label, text } from './RadioButton.css';

export type RadioButtonProps = React.InputHTMLAttributes<HTMLInputElement>;

export const RadioButton = forwardRef<HTMLInputElement, RadioButtonProps>(
  ({ children, disabled, ...props }, ref) => (
    <label className={label}>
      <input
        type="radio"
        disabled={disabled}
        className={field}
        ref={ref}
        {...props}
      />
      <span className={icon} />
      <span className={text}>{children}</span>
    </label>
  ),
);
