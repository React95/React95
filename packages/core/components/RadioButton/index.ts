import { RadioButton } from './RadioButton';

export default RadioButton;
