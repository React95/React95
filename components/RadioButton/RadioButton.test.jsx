import React from 'react';
import RadioButton from './RadioButton';
import { render, fireEvent } from 'react-testing-library';

describe('<RadioButton />', () => {
  describe('Snapshots', () => {
    it('should match snapshot without props', () => {
      const { container } = render(<RadioButton />);
      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with disabled prop', () => {
      const { container } = render(<RadioButton disabled />);
      expect(container).toMatchSnapshot();
    });
  });

  describe('onChange prop', () => {
    it('should call onChange callback when RadioButton is changed', () => {
      const onChangeMock = jest.fn();
      const { getByTestId } = render(
        <RadioButton onChange={onChangeMock} data-testid="radio">
          Radio
        </RadioButton>
      );

      fireEvent.click(getByTestId('radio'));

      expect(onChangeMock).toHaveBeenCalled();
    });
  });
});
