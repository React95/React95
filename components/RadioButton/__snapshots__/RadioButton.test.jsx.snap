// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioButton /> Snapshots should match snapshot with disabled prop 1`] = `
.c4 {
  width: 12px;
  height: 12px;
  content: '';
  display: inline-block;
  position: absolute;
  left: 0;
  top: 0;
  background-image: url(test-file-stub);
}

.c6 {
  padding: 1px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: absolute;
  top: 0;
  left: 18px;
}

.c2 {
  margin: 0;
  opacity: 0;
}

.c2:focus ~ .c5,
.c2:active + .c5 {
  border: 1px dotted;
  padding: 0;
}

.c2:checked + .c3 {
  background-image: url(test-file-stub);
}

.c2:disabled + .c3 {
  background-image: url(test-file-stub);
}

.c2:checked .c1:disabled + .c3 {
  background-image: url(test-file-stub);
}

.c0 {
  position: relative;
  margin-bottom: 10px;
  color: #868686;
  text-shadow: 0.5px 0.5px #d2d2d2;
}

<div>
  <label
    class="c0"
    disabled=""
  >
    <input
      class="c1 c2"
      disabled=""
      type="radio"
    />
    <span
      class="c3 c4"
    />
    <span
      class="c5 c6"
    />
  </label>
</div>
`;

exports[`<RadioButton /> Snapshots should match snapshot without props 1`] = `
.c4 {
  width: 12px;
  height: 12px;
  content: '';
  display: inline-block;
  position: absolute;
  left: 0;
  top: 0;
  background-image: url(test-file-stub);
}

.c6 {
  padding: 1px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: absolute;
  top: 0;
  left: 18px;
}

.c2 {
  margin: 0;
  opacity: 0;
}

.c2:focus ~ .c5,
.c2:active + .c5 {
  border: 1px dotted;
  padding: 0;
}

.c2:checked + .c3 {
  background-image: url(test-file-stub);
}

.c2:disabled + .c3 {
  background-image: url(test-file-stub);
}

.c2:checked .c1:disabled + .c3 {
  background-image: url(test-file-stub);
}

.c0 {
  position: relative;
  margin-bottom: 10px;
}

<div>
  <label
    class="c0"
  >
    <input
      class="c1 c2"
      type="radio"
    />
    <span
      class="c3 c4"
    />
    <span
      class="c5 c6"
    />
  </label>
</div>
`;
